## NumPy Fundamentals - Student Lab

## Lab Instructions

Complete all exercises below. Write your code in a Python file or Jupyter notebook. Test each task to ensure it produces the expected output.

**Assessment Criteria:**
- Code correctness and functionality
- Proper use of NumPy functions
- Code efficiency (use vectorized operations)
- Clean, readable code with appropriate comments
---

## Exercise 1: Array Creation and Attributes (15 points)

## Task 1.1: Create arrays using different methods (10 points)

Write code to create the following arrays:

 1. A 1D array containing values from 0 to 9
 2. A 2D array (3x3) containing values from 1 to 9
 3. A 3x4 array filled with zeros
 4. A 2x5 array filled with ones (data type: float32)
 5. A 4x4 identity matrix
 6. An array with values from 10 to 50 with step size 5
 7. An array with 8 evenly spaced values between 0 and 1

**Your Code:**

In [1]:
# Solution

import numpy as np

# 1.1D array containing values from 0 to 9
array_1D = np.arange(10)  # generates numbers from 0 to 
print("1D array (0 to 9):\n",array_1D,"\n")

# 2.2D array [3x3] containing values from 1 to 9
array_2D = np.arange(1,10).reshape(3, 3)  # creates values from 1 to 9 and arranges them into a 3x3 array
print("2D array [3x3] with values 1 to 9:\n",array_2D,"\n")

# 3.[3x4] array filled with zeros
zeros_array = np.zeros((3,4))  # creates a [3x4] array filled with zeros
print("[3x4] array filled with zeros:\n",zeros_array,"\n")

# 4.[2x5] array filled with ones (float32)
ones_array = np.ones((2,5),dtype=np.float32)  # creates a [2x5] array filled with ones in float(datatype as specified) format
print("[2x5] array filled with ones (float32):\n",ones_array,"\n")

# 5.[4x4] identity matrix
identity_matrix = np.eye(4)  # creates a 4x4 identity matrix
print("[4x4] identity matrix:\n",identity_matrix,"\n")

# 6.Array with values from 10 to 50 with step size 5
step_array = np.arange(10,51,5)  # creates values from 10 to 50 with a step size of 5
print("Array from 10 to 50 with step 5:\n",step_array,"\n")

# 7.Array with 8 evenly spaced values between 0 and 1
spaced_array = np.linspace(0,1,8)    # creates 8 evenly spaced values between 0 and 1 
# Referred ChatGPT for linspace syntax
print("Array with 8 evenly spaced values between 0 and 1:\n",spaced_array)

1D array (0 to 9):
 [0 1 2 3 4 5 6 7 8 9] 

2D array [3x3] with values 1 to 9:
 [[1 2 3]
 [4 5 6]
 [7 8 9]] 

[3x4] array filled with zeros:
 [[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]] 

[2x5] array filled with ones (float32):
 [[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]] 

[4x4] identity matrix:
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]] 

Array from 10 to 50 with step 5:
 [10 15 20 25 30 35 40 45 50] 

Array with 8 evenly spaced values between 0 and 1:
 [0.         0.14285714 0.28571429 0.42857143 0.57142857 0.71428571
 0.85714286 1.        ]


## Task 1.2: Explore array attributes (5 points)

Create a 3x4 array with values from 1 to 12. Print the following attributes:
- Shape
- Number of dimensions
- Total number of elements
- Data type
- Item size in bytes
- Total bytes consumed

**Your Code:**

In [2]:
# Solution

import numpy as np

# Creates a [3x4] array with values from 1 to 12
array = np.arange(1,13).reshape(3,4)  # Creating numbers from 1 to 12 and reshaping them into a [3x4] array

print("Array:\n",array,"\n")

# Shows Array attributes
print("Shape:",array.shape)   # array.shape gives the size of the array in terms of rows and columns               
print("Number of dimensions:",array.ndim)   # array.ndim tells how many dimensions the array has
print("Total number of elements:",array.size)  # array.size gives the total number of elements present in the array
print("Data type:",array.dtype)   # array.dtype shows the data type of elements stored in the array
print("Item size (in bytes):",array.itemsize)   # array.itemsize gives the size of one element in bytes
print("Total bytes consumed:",array.nbytes)   # array.nbytes gives the total memory used by the array in bytes


Array:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]] 

Shape: (3, 4)
Number of dimensions: 2
Total number of elements: 12
Data type: int64
Item size (in bytes): 8
Total bytes consumed: 96


## Exercise 2: Indexing and Slicing (20 points)

## Task 2.1: Basic indexing (5 points)

Given the array: `arr = np.array([10, 20, 30, 40, 50, 60, 70, 80, 90, 100])`

Write code to:
1. Access the first element
2. Access the last element using negative indexing
3. Access the 5th element

Then create a 2D array with shape (3, 4) and:
- Access the element at row 1, column 2
- Access the element at row 0, column 3
- Access the last element

**Your Code:**

In [3]:
# Solution 

import numpy as np

# Given array
array = np.array([10,20,30,40,50,60,70,80,90,100]) # Creating an 1D array with given values

# 1. Access the first element
print("First element:", array[0])  # Accessing the first element of the 1D array using index 0
# 2.Access the last element using negative indexing
print("Last element:", array[-1])   # Accessing the last element using negative indexing
# 3. Access the 5th element
print("5th element:", array[4])   # Accessing the 5th element by using index 4       

print()

# Then create a 2D array with shape (3, 4) and:
array_2D = np.arange(1,13).reshape(3,4)  # Creating a 2D array of size 3x4 using values from 1 to 12
print("2D Array:\n",array_2D,"\n")

# Access the element at row 1, column 2
print("Element at row 1, column 2:", array_2D[1, 2])  # Accessing the element at row index 1 and column index 2
# Access the element at row 0, column 3
print("Element at row 0, column 3:", array_2D[0, 3])  # Accessing the element at row index 0 and column index 3
# Access the last element
print("Last element:", array_2D[-1, -1])  # Accessing the last element of the 2D array using negative indexing

First element: 10
Last element: 100
5th element: 50

2D Array:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]] 

Element at row 1, column 2: 7
Element at row 0, column 3: 4
Last element: 12


## Task 2.2: Slicing operations (8 points)

Given `arr = np.arange(0, 20)`:

1. Extract elements from index 5 to 15 (inclusive)
2. Extract every 3rd element from the array
3. Reverse the entire array using slicing

Create a 4x4 2D array and:
- Extract the first 2 rows and first 3 columns
- Extract the last 2 rows and last 2 columns
- Extract every other row with all columns

**Your Code:**

In [4]:
# Solution 

import numpy as np

# Given array
array = np.arange(0,20)  # Creating a 1D array with values from 0 to 19
print("Original array:\n",array,"\n")

# 1. Extract elements from index 5 to 15 (inclusive)
slice_1 = array[5:16]   # Extracting elements starting from index 5 up to index 15 using slicing
print("Elements from index 5 to 15:",slice_1)

# 2. Extract every 3rd element from the array
slice_2 = array[::3]    # Extracting every 3rd element from the array using step slicing
print("Every 3rd element:",slice_2)

# 3. Reverse the entire array using slicing
slice_3 = array[::-1]   # Reversing the entire array using slicing
print("Reversed array:",slice_3)

print("\n-----------------------------\n")

# Create a [4x4] 2D array and:
array_2D = np.arange(1,17).reshape(4,4)  # Creating a [4x4] 2D array using values from 1 to 16
print("2D Array:\n",array_2D,"\n")

#  Extract the first 2 rows and first 3 columns
sub_array_1 = array_2D[:2, :3]     # Extracting the first 2 rows and first 3 columns from the 2D array
print("First 2 rows and first 3 columns:\n", sub_array_1)

#  Extract the last 2 rows and last 2 columns
sub_array_2 = array_2D[-2:, -2:]     # Extracting the last 2 rows and last 2 columns using negative indexing
print("\nLast 2 rows and last 2 columns:\n", sub_array_2)

# Extract every other row with all columns
sub_array_3 = array_2D[::2, :]     # Extracting every alternate row while keeping all columns
print("\nEvery other row:\n", sub_array_3)
#Referred chatgpt for syntaxes

Original array:
 [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19] 

Elements from index 5 to 15: [ 5  6  7  8  9 10 11 12 13 14 15]
Every 3rd element: [ 0  3  6  9 12 15 18]
Reversed array: [19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2  1  0]

-----------------------------

2D Array:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]] 

First 2 rows and first 3 columns:
 [[1 2 3]
 [5 6 7]]

Last 2 rows and last 2 columns:
 [[11 12]
 [15 16]]

Every other row:
 [[ 1  2  3  4]
 [ 9 10 11 12]]


## Task 2.3: Boolean and fancy indexing (7 points)

Given `arr = np.array([15, 23, 8, 42, 11, 37, 19, 5, 33, 28])`:

1. Use boolean indexing to extract all values greater than 20
2. Extract all even numbers
3. Extract values between 10 and 30 (inclusive)
4. Use fancy indexing to extract elements at indices 0, 3, 5, and 8

**Your Code:**

In [5]:
# Solution

import numpy as np

# Given array
given_array = np.array([15,23,8,42,11,37,19,5,33,28])  # Creating a NumPy array with the given values
print("Original array:")
print(given_array)

# 1. Use boolean indexing to extract all values greater than 20
values_greater_than_20 = given_array[given_array>20]  # condition to extract values greater than 20
print("Values greater than 20 are:",values_greater_than_20)

# 2. Extract all even numbers
even_numbers = given_array[given_array%2==0]  # condition to extract only even numbers from the array
print("Even numbers are:",even_numbers)

# 3. Extract values between 10 and 30 (inclusive)
values_between_10_30 = given_array[(given_array>=10)&(given_array<=30)]  # condition to extract values that lie between 10 and 30 (inclusive)
print("Values between 10 and 30(inclusive) are:",values_between_10_30)

# 4. Use fancy indexing to extract elements at indices 0, 3, 5, and 8
fancy_indexing = given_array[[0,3,5,8]]   # Using fancy indexing to extract elements from specific indices
print("Elements at indices 0,3,5,and 8 are:",fancy_indexing)


Original array:
[15 23  8 42 11 37 19  5 33 28]
Values greater than 20 are: [23 42 37 33 28]
Even numbers are: [ 8 42 28]
Values between 10 and 30(inclusive) are: [15 23 11 19 28]
Elements at indices 0,3,5,and 8 are: [15 42 37 33]


## Exercise 3: Array Operations (15 points)

### Task 3.1: Arithmetic operations (8 points)

Given two arrays:
- `arr1 = np.array([10, 20, 30, 40, 50])`
- `arr2 = np.array([5, 10, 15, 20, 25])`

Perform the following operations:
1. Element-wise addition
2. Element-wise subtraction
3. Element-wise multiplication
4. Element-wise division
5. Square each element of arr1

Then perform scalar operations on arr1:
- Add 100 to all elements
- Multiply all elements by 3
- Divide all elements by 10

**Your Code:**

In [6]:
# Solution

import numpy as np

# Given arrays
array_1 = np.array([10,20,30,40,50])  # Creating two NumPy arrays with given values
array_2 = np.array([5,10,15,20,25])

print("Array 1:",array_1)    
print("Array 2:",array_2,"\n")

# 1. Element-wise addition
addition = array_1+array_2   # Adding the elements of array_1 and array_2
print("Element-wise addition:",addition)

# 2. Element-wise subtraction
subtraction = array_1-array_2  # Subtracting the elemnts of array_1 and array_2
print("Element-wise subtraction:",subtraction)

# 3. Element-wise multiplication
multiplication = array_1*array_2   # Multiplying the elements of array_1 and array_2
print("Element-wise multiplication:",multiplication)

# 4. Element-wise division
division = array_1/array_2  # Dividing the elements array_1 and array_2
print("Element-wise division:",division)

# Square each element of array_1
squares = array_1** 2   # Squaring each element of the array_1
print("Square of elements in Array 1:",squares)

print("\n--- Scalar Operations on Array 1 ---")

# Scalar operations
# Add 100 to all elements
scalar_addition = array_1+100    # Adding 100 to all elements of the array_1
print("Add 100 to all elements:",scalar_addition)

# Multiply all elements by 3
scalar_multiplication = array_1*3   # Multiplying all elements of the array_1 by 3
print("Multiply all elements by 3:",scalar_multiplication)

# Divide all elements by 10
scalar_division = array_1/10       # Dividing all elements of the array_1 by 10
print("Divide all elements by 10:",scalar_division)

Array 1: [10 20 30 40 50]
Array 2: [ 5 10 15 20 25] 

Element-wise addition: [15 30 45 60 75]
Element-wise subtraction: [ 5 10 15 20 25]
Element-wise multiplication: [  50  200  450  800 1250]
Element-wise division: [2. 2. 2. 2. 2.]
Square of elements in Array 1: [ 100  400  900 1600 2500]

--- Scalar Operations on Array 1 ---
Add 100 to all elements: [110 120 130 140 150]
Multiply all elements by 3: [ 30  60  90 120 150]
Divide all elements by 10: [1. 2. 3. 4. 5.]


### Task 3.2: Universal functions (7 points)

Given `arr = np.array([1, 4, 9, 16, 25, 36])`:

1. Calculate square root of all elements
2. Calculate exponential (e^x) of all elements
3. Calculate natural logarithm of all elements

Create an array of angles `[0, π/6, π/4, π/3, π/2]` and:
- Calculate sine values
- Calculate cosine values

For the array `[1.23, 2.56, 3.89, 4.12]`:
- Round to nearest integer
- Apply floor operation
- Apply ceiling operation

**Your Code:**

In [7]:
# Solution

import numpy as np

# Given array
array = np.array([1,4,9,16,25,36])   # Creating a NumPy array with perfect square values
print("Original array:",array,"\n")

# 1. Calculate square root of all elements
sqrt_array = np.sqrt(array)  # np.sqrt() calculates the square root of each element in the array
print("Square root:",sqrt_array)

# 2. Calculate exponential (e^x) of all elements
exponential_array = np.exp(array)      # np.exp() calculates e raised to the power of each element in the array
print("Exponential (e^x):",exponential_array)

# 3. Calculate natural logarithm of all elements
log_array = np.log(array)  # np.log() calculates the natural logarithm of each element in the array
print("Natural logarithm:",log_array)

print("\n-----------------------------\n")

# Create an array of angles `[0, π/6, π/4, π/3, π/2]`
angles = np.array([0,np.pi/6,np.pi/4,np.pi/3,np.pi/2])   
print("Angles:",angles,"\n")

# Calculate sine values
sin_values = np.sin(angles)  # np.sin() calculates the sine value for each angle in the array
print("Sine values:",sin_values)

# Calculate cosine values
cos_values = np.cos(angles)  # np.cos() calculates the cosine value for each angle in the array
print("Cosine values:",cos_values)

print("\n-----------------------------\n")

# Given float array
float_array = np.array([1.23,2.56,3.89,4.12])
print("Float array:",float_array,)

# Round to nearest integer
rounded = np.round(float_array) # np.round() rounds each value to the nearest integer
print("Rounded values:",rounded)

# Apply floor operation
floor_values = np.floor(float_array)  # np.floor() gives the greatest integer less than or equal to each value
print("Floor values:",floor_values)

# Apply ceiling operation
ceil_values = np.ceil(float_array)  # np.ceil() gives the smallest integer greater than or equal to each value
print("Ceiling values:",ceil_values)
# Reffered chatgpt for assistance with creating array on angles with radians and ceil() function

Original array: [ 1  4  9 16 25 36] 

Square root: [1. 2. 3. 4. 5. 6.]
Exponential (e^x): [2.71828183e+00 5.45981500e+01 8.10308393e+03 8.88611052e+06
 7.20048993e+10 4.31123155e+15]
Natural logarithm: [0.         1.38629436 2.19722458 2.77258872 3.21887582 3.58351894]

-----------------------------

Angles: [0.         0.52359878 0.78539816 1.04719755 1.57079633] 

Sine values: [0.         0.5        0.70710678 0.8660254  1.        ]
Cosine values: [1.00000000e+00 8.66025404e-01 7.07106781e-01 5.00000000e-01
 6.12323400e-17]

-----------------------------

Float array: [1.23 2.56 3.89 4.12]
Rounded values: [1. 3. 4. 4.]
Floor values: [1. 2. 3. 4.]
Ceiling values: [2. 3. 4. 5.]


## Exercise 4: Array Manipulation (20 points)

### Task 4.1: Reshaping and transposing (10 points)

Create an array with values from 1 to 12, then:
1. Reshape it to 3x4
2. Reshape it to 2x6
3. Reshape it to 2x2x3 (3D array)
4. Transpose the 3x4 array
5. Flatten the 3x4 array back to 1D

**Your Code:**

In [ ]:
# Solution

import numpy as np

# Creates an array with values from 1 to 12
array = np.arange(1,13)  # np.arange(1, 13) creates an array with values from 1 to 12
print("Original array:\n",array,"\n")

# 1.Reshape it to [3x4]
array_3x4 = array.reshape(3,4)  # array.reshape(3,4) rearranges the 12 elements into 3 rows and 4 columns
print("Reshaped to [3x4]:\n", array_3x4,"\n")

# 2.Reshape it to [2x6]
array_2x6 = array.reshape(2,6)   # array.reshape(2,6) rearranges the elements into 2 rows and 6 columns
print("Reshaped to [2x6]:\n",array_2x6,"\n")

# 3.Reshape it to [2x2x3] (3D array)
array_3D = array.reshape(2, 2, 3)  # array.reshape(2,2,3) converts the 1D array into a 3D array with 2 blocks, 2 rows, and 3 columns
print("Reshaped to [2x2x3] (3D array):\n",array_3D,"\n")

# 4. Transpose the [3x4] array
transposed = array_3x4.T   # .T transposes the 3x4 array by converting rows into columns and columns into rows
print("Transpose of 3x4 array:\n",transposed,"\n")

# 5. Flatten the [3x4] array back to 1D
flattened = array_3x4.flatten()     # flatten() converts the 2D array back into a 1D array by placing all elements in a single line
print("Flattened 1D array:\n",flattened)
# Referred chatgpt for assistance with Transpose() and flatten() functions

### Task 4.2: Stacking and splitting (10 points)

Create three 1D arrays: `[1, 2, 3]`, `[4, 5, 6]`, `[7, 8, 9]`

1. Stack them vertically to create a 3x3 array
2. Stack them horizontally to create a 1D array

Create an array from 1 to 12 and:
- Split it into 4 equal parts

Create a 4x4 array with values 1-16 and:
- Split it horizontally into 2 parts

**Your Code:**

In [ ]:
# Solution

import numpy as np

# Create three 1D arrays
a = np.array([1,2,3])  # Creating three 1D arrays 
b = np.array([4,5,6])
c = np.array([7,8,9])

print("Array a:",a)
print("Array b:",b)
print("Array c:",c,"\n")

# 1. Stack them vertically to create a 3x3 array
vertical_stack = np.vstack((a,b,c))  # np.vstack() stacks the arrays one below the other to form a 3x3 matrix
print("Vertically stacked array:\n",vertical_stack,"\n")

# 2. Stack them horizontally to create a 1D array
horizontal_stack = np.hstack((a,b,c))  # np.hstack() joins the arrays side by side to form a single 1D array
print("Horizontally stacked array:",horizontal_stack,"\n")

print("-----------------------------\n")

# Create an array from 1 to 12 
array = np.arange(1,13)  # np.arange(1, 13) creates a 1D array with values from 1 to 12
print("Array from 1 to 12:",array,"\n")

# Split into 4 equal parts
split_1D = np.split(array,4)    # np.split(array, 4) divides the array into 4 equal-sized sub-arrays
print("Split into 4 equal parts:")
for part in split_1D:
    print(part)

print("\n-----------------------------\n")

# Create a [4x4] array with values 1 to 16
array_2D = np.arange(1,17).reshape(4,4)  # np.arrange(1, 17) creates an array with values from 1 to 16 and reshape(4, 4) changes it to a 4x4 matrix
print("[4x4] Array:\n",array_2D,"\n")

# Split it horizontally into 2 parts
split_2D = np.hsplit(array_2D,2)   # np.hsplit() splits the 2D array horizontally into 2 equal parts based on columns   
print("Horizontally split into 2 parts:")
for part in split_2D:
    print(part,"\n")
# Used chatgpt for assistance with split() and hsplit() functions

## Exercise 5: Statistical Operations (15 points)

### Task 5.1: Basic statistics (7 points)

Given `arr = np.array([12, 45, 23, 67, 34, 89, 56, 78, 90, 31])`:

Calculate and print:
1. Sum of all elements
2. Mean (average)
3. Median
4. Standard deviation
5. Variance
6. Minimum value
7. Maximum value
8. Index of minimum value
9. Index of maximum value

**Your Code:**

In [ ]:
# Solution

import numpy as np

# Given array
array = np.array([12,45,23,67,34,89,56,78,90,31])  # Creating a NumPy array with the given set of values
print("Array:",array,"\n")

# 1.Sum of all elements
total_sum = np.sum(array) # np.sum() calculates the total of all elements present in the array
print("Sum:",total_sum)

# 2.Mean(average)
mean_value = np.mean(array)  # np.mean() calculates the average value of all elements in the array
print("Mean:",mean_value)

# 3.Median
median_value = np.median(array)   # np.median() finds the middle value after arranging the elements in order
print("Median:",median_value)

# 4.Standard deviation
standard_value = np.std(array)     # np.std() calculates how much the values in the array deviate from the mean
print("Standard deviation:",standard_value)

# 5.Variance
variance_value = np.var(array)   # np.var() calculates the variance, which is the square of the standard deviation
print("Variance:",variance_value)

# 6. Minimum value
minimum_value = np.min(array)    # np.min() finds the smallest value present in the array
print("Minimum value:",minimum_value)

# 7. Maximum value
maximum_value = np.max(array)    # np.max() finds the largest value present in the array
print("Maximum value:",maximum_value)

# 8. Index of minimum value
minimum_index = np.argmin(array)    # np.argmin() gives the index position of the minimum value in the array 
print("Index of minimum value:",minimum_index)

# 9. Index of maximum value
maximum_index = np.argmax(array)     # np.argmax() gives the index position of the maximum value in the array
print("Index of maximum value:",maximum_index)
# Used chatgpt for assistance with std(),argmin() and argmax() functions

### Task 5.2: Axis-based statistics (8 points)

Create a 3x3 array:
```
[[10, 20, 30],
 [40, 50, 60],
 [70, 80, 90]]
```

Calculate:
1. Sum of all elements
2. Sum along axis 0 (column-wise sum)
3. Sum along axis 1 (row-wise sum)
4. Mean along axis 0
5. Mean along axis 1
6. Maximum values along axis 0
7. Maximum values along axis 1

**Your Code:**

In [ ]:
# Solution

import numpy as np

# Create a [3x3] array
array = np.array([[10, 20, 30],    # Creating a 3x3 NumPy array with given values
                [40, 50, 60],
                [70, 80, 90]])
print("Array:\n",array,"\n")

# 1.Sum of all elements
total_sum = np.sum(array)   # np.sum() adds all the elements present in the entire array
print("Sum of all elements:",total_sum)

# 2.Sum along axis 0 (column-wise)
column_sum = np.sum(array,axis=0)   # np.sum(array, axis=0) adds elements column-wise, so each column is summed separately
print("Column-wise sum (axis 0):",column_sum)

# 3.Sum along axis 1 (row-wise)
row_sum = np.sum(array,axis=1) # np.sum(array, axis=1) adds elements row-wise, so each row is summed separately
print("Row-wise sum (axis 1):",row_sum)

# 4.Mean along axis 0 (column-wise)
column_mean = np.mean(array,axis=0)  # np.mean(array, axis=0) calculates the average of each column
print("Column-wise mean (axis 0):",column_mean)

# 5.Mean along axis 1 (row-wise) 
row_mean = np.mean(array,axis=1)   # np.mean(array, axis=1) calculates the average of each row
print("Row-wise mean (axis 1):",row_mean)

# 6.Maximum values along axis 0 
column_maximum = np.max(array,axis=0)   # np.max(array, axis=0) finds the maximum value in each column
print("Column-wise maximum (axis 0):",column_maximum)

# 7.Maximum values along axis 1 
row_maximum = np.max(array,axis=1)   # np.max(array, axis=1) finds the maximum value in each row
print("Row-wise maximum (axis 1):",row_maximum)
# Used chatgpt for assistance with axis=0 and axis=1 instances usage

## Exercise 6: Real-World Application - Student Grade Analysis (25 points)

You are given grade data for 5 students across 4 subjects (Math, Science, English, History):

```python
grades = np.array([
    [85, 90, 78, 88],
    [92, 88, 95, 90],
    [76, 82, 80, 75],
    [88, 85, 90, 92],
    [95, 92, 88, 94]
])
```

**Tasks:**

1. **Calculate the average grade for each student** (5 points)
   - Output should show Student 1: X.XX, Student 2: Y.YY, etc.

2. **Calculate the average grade for each subject** (5 points)
   - Output: Math: X.XX, Science: Y.YY, etc.

3. **Find the highest and lowest grades in each subject** (5 points)
   - Output for each subject: Highest = X, Lowest = Y

4. **Identify students with an average grade above 85** (3 points)
   - Print student number and their average

5. **Calculate the overall class average** (2 points)
   - Average of all grades across all students

6. **Normalize all grades to a scale of 0-10** (5 points)
   - Use min-max normalization formula: (value - min) / (max - min) * 10
   - Print the normalized grades array

**Bonus:** Identify the top-performing student (+2 points)
``
**Expected Output Format:**
```
--- Task 1: Student Averages ---
Student 1: XX.XX
...
--- Task 2: Subject Averages ---
Math: XX.XX
...
(Continue for all tasks)
```
---
**Your Code:**

In [ ]:
# Solution

import numpy as np

#Given Grade Data
grades = np.array([        # Creating a 2D array 
    [85,90,78,88],   
    [92,88,95,90],
    [76,82,80,75],
    [88,85,90,92],
    [95,92,88,94]
])

subject_names = ["Math","Science","English","History"]   # Storing subject names separately 


# 1.Calculate the average grade for each student
print("--- Task 1: Student Averages ---")
student_average = np.mean(grades,axis=1)   # np.mean(grades, axis=1) calculates the average marks for each student by averaging across columns
for i, avg in enumerate(student_average,1): # enumerate(student_average, 1) is used to number students starting from 1 while printing their averages
    print(f"Student {i}: {avg:.2f}")

# 2.Calculate the average grade for each subject
print("\n--- Task 2: Subject Averages ---")
subject_average = np.mean(grades,axis=0)     # np.mean(grades, axis=0) calculates the average marks for each subject by averaging across rows
for name,avg in zip(subject_names,subject_average):  # zip() is used to pair each subject name with its corresponding average value
    print(f"{name}:{avg:.2f}")

# 3.Find the highest and lowest grades in each subject
print("\n--- Task 3: Highest and Lowest Grades per Subject ---")
subject_maximum = np.max(grades,axis=0)  # np.max(grades, axis=0) finds the highest score obtained in each subject
subject_minimum = np.min(grades,axis=0)  # np.min(grades, axis=0) finds the lowest score obtained in each subject
for name,mx,mn in zip(subject_names,subject_maximum,subject_minimum): 
    print(f"{name}:Highest={mx},Lowest={mn}")

# 4.Identify students with an average grade above 85
print("\n--- Task 4: Students with Average Above 85 ---")
for i,avg in enumerate(student_average,1):    # condition to check which students have an average grade greater than 85
    if avg>85:
        print(f"Student{i}:{avg:.2f}")
        
# 5.Calculate the overall class average
print("\n--- Task 5: Overall Class Average ---")
class_average = np.mean(grades)  # np.mean(grades) calculates the overall class average by considering all grades together
print(f"Overall Class Average: {class_average:.2f}")

# 6.Normalize all grades to a scale of 0-10
print("\n--- Task 6: Normalized Grades (0-10 scale) ---")
minimum_grade = np.min(grades)  # np.min(grades) finds the lowest grade out of all the students
maximum_grade = np.max(grades)  # np.max(grades) finds the highest grade out of all the students
normalized_grades = (grades-minimum_grade)/(maximum_grade-minimum_grade)*10   # condition for normalizing grades
print(normalized_grades)

# Identify the top-performing student 
top_student_index = np.argmax(student_average)  # np.argmax(student_average) finds the index of the student with the highest average score
print(f"\n--- Top-Performing Student ---")
print(f"Student {top_student_index + 1} with average grade: {student_average[top_student_index]:.2f}")
# Used chatgpt for assistance with enumerate() and zip() functions

## Exercise 7: Challenge - Sales Data Analysis (30 points)

# You are analyzing sales data for a company with 4 products over 6 months:

```python
sales_data = np.array([
    [120, 135, 150, 145, 160, 175],  # Product A
    [90, 95, 100, 110, 115, 120],    # Product B
    [150, 145, 155, 160, 170, 180],  # Product C
    [80, 85, 82, 88, 92, 95]         # Product D
])
```

**Tasks:**

1. **Calculate total sales for each product** (5 points)
   - Sum across all months for each product

2. **Calculate total sales for each month** (5 points)
   - Sum across all products for each month (Jan to Jun)

3. **Identify the month with the highest total sales** (4 points)
   - Print the month name and total sales

4. **Calculate the growth rate for each product** (6 points)
   - Growth rate = ((Last month - First month) / First month) × 100
   - Print as percentage for each product

5. **Find products with average monthly sales greater than 120k** (4 points)
   - Calculate average for each product
   - Print products meeting the criteria

6. **Calculate quarter-wise sales** (6 points)
   - Q1: January-March
   - Q2: April-June
   - Print sales for each product in each quarter

**Bonus:** Predict next month (July) sales for each product (+5 points)
- Use a simple method: average of last 2 months + average monthly growth

**Your Code:**

In [ ]:
# Solution

import numpy as np

sales_data = np.array([      # Creating a 2D array
    [120,135,150,145,160,175],  # Product A  
    [90,95,100,110,115,120],    # Product B
    [150,145,155,160,170,180],  # Product C
    [80,85,82,88,92,95]         # Product D
])

product_names = ["Product A","Product B","Product C","Product D"]  # Storing product names 
month_names = ["Jan","Feb","Mar","Apr","May","Jun"]  # Storing month names

# 1.Calculate total sales for each product
print("--- Task 1: Total Sales per Product ---")
total_product_sales = np.sum(sales_data,axis=1)  # np.sum(sales_data, axis=1) adds monthly sales row-wise to get total sales for each product
for name, total in zip(product_names,total_product_sales):   # zip() is used to print each product name along with its total sales
    print(f"{name}: {total}k")          
    
# 2.Calculate total sales for each month
print("\n--- Task 2: Total Sales per Month ---")
total_month_sales = np.sum(sales_data,axis=0)  # np.sum(sales_data, axis=0) adds sales column-wise to get total sales for each month
for month, total in zip(month_names,total_month_sales):
    print(f"{month}: {total}k")

# 3.Identify the month with the highest total sales
print("\n--- Task 3: Month with Highest Total Sales ---")
max_month_index = np.argmax(total_month_sales)  # np.argmax() finds the index of the month with the highest total sales
print(f"Highest sales in {month_names[max_month_index]}: {total_month_sales[max_month_index]}k") # Using the index to display the corresponding month name and sales value 

# 4.Calculate the growth rate for each product
print("\n--- Task 4: Growth Rate per Product ---")
growth_rate = ((sales_data[:, -1]-sales_data[:, 0])/sales_data[:, 0])*100  # sales_data[:, -1] selects the last month’s sales and sales_data[:, 0] selects the first month’s sales
for name, gr in zip(product_names,growth_rate):
    print(f"{name}: {gr:.2f}%")

# 5.Find products with average monthly sales greater than 120k
print("\n--- Task 5: Products with Average Monthly Sales > 120k ---")
average_sales = np.mean(sales_data,axis=1)  # np.mean(sales_data, axis=1) calculates the average monthly sales for each product.
for name, avg in zip(product_names,average_sales):  # condition to check products with avg mothly sales > 120K.
    if avg>120:
        print(f"{name}: {avg:.2f}k")

# 6.Calculate quarter-wise sales
print("\n--- Task 6: Quarter-wise Sales ---")
# Q1: Jan-Mar
quarter1_sales = np.sum(sales_data[:,0:3],axis=1)  # sales_data[:, 0:3] selects sales from Jan to Mar for Quarter 1.
# Q2: Apr-Jun
quarter2_sales = np.sum(sales_data[:,3:6],axis=1)  # sales_data[:, 3:6] selects sales from Apr to Jun for Quarter 2.
for name,q1,q2 in zip(product_names,quarter1_sales,quarter2_sales):
    print(f"{name}: Quarter1 = {q1}k,Quarter2 = {q2}k")       

# 
print("\n--- Predicting July Sales ---")
# Simple prediction:average of last 2 months + average monthly growth.
last_two_average = np.mean(sales_data[:,-2:],axis=1)  # np.mean(sales_data, axis=1) calculates the average of the last two months’ sales.
monthly_growth = (sales_data[:,-1]-sales_data[:,0])/(sales_data.shape[1]-1) # sales_data.shape[1] gives the total number of months in the dataset.
july_prediction = last_two_average+monthly_growth  # Calculating average monthly growth based on overall sales increase.
for name,pred in zip(product_names,july_prediction):  # Predicting July sales by adding recent average sales and monthly growth.
    print(f"{name}: Predicted July sales ={pred:.2f}k")
# Used chatgpt for assistance with growth_rate condition,quarter1_sales condition,quarter2_sales condition and for july_prediction condition.  

## Submission Guidelines

1. Complete all exercises in order
2. Test your code to ensure it runs without errors
3. Include comments explaining your logic for complex operations
4. Format your output clearly with appropriate labels
5. Save your file as `numpy_lab_yourname.py` or `numpy_lab_yourname.ipynb`

---

## Evaluation Rubric

| Exercise | Points | Criteria |
|----------|--------|----------|
| Exercise 1 | 15 | Array creation and attributes |
| Exercise 2 | 20 | Indexing and slicing techniques |
| Exercise 3 | 15 | Array operations and ufuncs |
| Exercise 4 | 20 | Array manipulation |
| Exercise 5 | 15 | Statistical operations |
| Exercise 6 | 25 | Real-world application |
| Exercise 7 | 30 | Challenge problem |
| **Total** | **140** | |

**Bonus Points:** Up to +7 additional points available

**Grading Scale:**
- A: 126-140+ points (90%+)
- B: 112-125 points (80-89%)
- C: 98-111 points (70-79%)
- D: 84-97 points (60-69%)
- F: Below 84 points (Below 60%)

---

## Tips for Success

1. **Use vectorized operations** instead of loops whenever possible
2. **Understand the axis parameter**: axis=0 (down columns), axis=1 (across rows)
3. **Check array shapes** frequently to avoid dimension errors
4. **Use meaningful variable names** for clarity
5. **Test incrementally** - don't write all code at once
6. **Read NumPy documentation** when stuck: https://numpy.org/doc/

Good luck!